jax/interpreters/xla.py

```diff
@@ -516,22 +534,88 @@ def _xla_callable(fun, device, backend, name, *arg_specs):
       extend_name_stack(wrap_name(name, 'jit')), *xla_args)
   built = c.Build(c.Tuple(*out_nodes))
 
+  if device:
+    device_assignment = ((device.id,),)
+  else:
+    device_assignment = xb.get_backend(backend).get_default_device_assignment(nreps, nparts)
+    device_assignment = onp.vectorize(lambda d: d.id)(onp.array(device_assignment))
+
   options = xb.get_compile_options(
-      num_replicas=nreps,
-      num_partitions=1,
-      device_assignment=(device.id,) if device else None)
+      num_replicas=nreps, num_partitions=nparts, device_assignment=device_assignment)
   compiled = built.Compile(compile_options=options, backend=xb.get_backend(backend))
 
-  if nreps == 1:
-    return partial(_execute_compiled, compiled, backend, result_handlers, tuple_args)
-  else:
-    return partial(_execute_replicated, compiled, backend, result_handlers, tuple_args)
+  return create_execute_fn(
+      compiled, device_assignment=device_assignment, backend=backend,
+      tuple_args=tuple_args, result_handlers=result_handlers,
+      out_partition_ids=out_partition_ids)
+
+def create_execute_fn(
+    compiled, *, device_assignment, backend, tuple_args, result_handlers, out_partition_ids):
+  # TODO(cjfj): Expose device assignment from the executable.
+  local_devices = compiled.local_devices()
+  # We are assuming here that `local_devices` are ordered by replica then partition.
+  # TODO(cjfj): Expose logical devices from the executable, rather than re-computing here.
```

Comment:
This is now done in https://github.com/tensorflow/tensorflow/commit/1e5dc679bbc3f2ee1941218ab184dafabafacf95. If we bump the jaxlib version, I can remove this code.